iot_packet_verifier/src/solana.rs

```diff
@@ -0,0 +1,274 @@
+    }
+
+    async fn burn_data_credits(
+        &self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+    ) -> Result<(), Self::Error> {
+        tracing::info!("Burning {} DC from {}", amount, payer);
+
+        // Fetch the sub dao epoch info:
+        const EPOCH_LENGTH: u64 = 60 * 60 * 24;
+        let epoch = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)?
+            .as_secs()
+            / EPOCH_LENGTH;
+        let (sub_dao_epoch_info, _) = Pubkey::find_program_address(
+            &[
+                "sub_dao_epoch_info".as_bytes(),
+                self.program_cache.sub_dao.as_ref(),
+                &epoch.to_le_bytes(),
+            ],
+            &helium_sub_daos::ID,
+        );
+
+        // Fetch escrow account
+        let ddc_key = pdas::delegated_data_credits(&self.program_cache.sub_dao, payer);
+        let account_data = self.provider.get_account_data(&ddc_key).await?;
+        let mut account_data = account_data.as_ref();
+        let escrow_account =
+            DelegatedDataCreditsV0::try_deserialize(&mut account_data)?.escrow_account;
+
+        let instructions = {
+            let request = RequestBuilder::from(
+                data_credits::id(),
+                &self.cluster,
+                std::rc::Rc::new(Keypair::from_bytes(&self.keypair).unwrap()),
+                Some(CommitmentConfig::confirmed()),
+                RequestNamespace::Global,
+            );
+
+            let accounts = accounts::BurnDelegatedDataCreditsV0 {
+                sub_dao_epoch_info,
+                dao: self.program_cache.dao,
+                sub_dao: self.program_cache.sub_dao,
+                account_payer: self.program_cache.account_payer,
+                data_credits: self.program_cache.data_credits,
+                delegated_data_credits: pdas::delegated_data_credits(
+                    &self.program_cache.sub_dao,
+                    payer,
+                ),
+                token_program: spl_token::id(),
+                helium_sub_daos_program: helium_sub_daos::id(),
+                system_program: solana_program::system_program::id(),
+                dc_burn_authority: self.program_cache.dc_burn_authority,
+                dc_mint: self.program_cache.dc_mint,
+                escrow_account,
+                registrar: self.program_cache.registrar,
+            };
+            let args = instruction::BurnDelegatedDataCreditsV0 {
+                args: data_credits::BurnDelegatedDataCreditsArgsV0 { amount },
+            };
+
+            // As far as I can tell, the instructions function does not actually have any
+            // error paths.
+            request
+                .accounts(accounts)
+                .args(args)
+                .instructions()
+                .unwrap()
+        };
+
+        let blockhash = self.provider.get_latest_blockhash().await?;
+        let signer = Keypair::from_bytes(&self.keypair).unwrap();
```

Comment:
is the logic behind the unwrap here that we've already validated the keypair can be constructed from the byte array when the Settings are created so we don't expect that byte array to change in-memory after the service has started up and gotten beyond that step?

Reply:
That is correct